src/v/serde/serde.h

```diff
@@ -344,6 +361,37 @@ void write(iobuf& out, T t) {
 
         write(out, t.is_ipv4());
         write(out, std::move(address_bytes));
+    } else if constexpr (is_chrono_duration<Type>) {
+        // We explicitly serialize it as ns to avoid any surprises like
+        // seastar updating underlying duration types without
+        // notice. See https://github.com/redpanda-data/redpanda/pull/5002
+        //
+        // Check for overflows, chrono type casts are prone to overflows.
+        // For ex: a millisecond and nanosecond use the same underlying
+        // type int64_t but converting from one to other can easily overflow,
+        // this is by design.
+        // Since we serialize with ns precision, there is a restriction of
+        // nanoseconds::max()'s equivalent on the duration to be serialized.
+        // On a typical platform which uses int64_t for 'rep', it roughly
+        // translates to ~292 years.
+        if (unlikely(check_duration_overflow(t))) {
```

Comment:
Can we remove the floating point arithmetic here? Can we place restrictions on what serde supports to make this more straight forward? Can we use our own sentinel values for min/max representation? Should all callers have to provide nanoseconds and figure it out for themselves?

Reply:
> Can we remove the floating point arithmetic here

I tried an alternative way of detecting overflows by casting to ns and back and see if we have the same value but that doesn't work.

> Can we use our own sentinel values for min/max representation? Should all callers have to provide nanoseconds and figure it out for themselves?

We can't control the `Rep` of duration to use sentinels? This code works with any custom duration type and forcing the callers to pass nanoseconds means we are just forcing them to do the same check and serialization is not seamless.